modules/pam_unix/passverify.c

```diff
@@ -395,6 +395,14 @@ PAMH_ARG_DECL(char * create_password_hash,
 	if (on(UNIX_MD5_PASS, ctrl)) {
 		/* algoid = "$1" */
 		return crypt_md5_wrapper(password);
+#if (defined(CRYPT_PREFERRED_METHOD_AVAILABLE) && CRYPT_PREFERRED_METHOD_AVAILABLE)
+	} else if (on(UNIX_CRYPT_DEFAULT_PASS, ctrl)
+		  && crypt_preferred_method() != NULL) {
```

Comment:
In which case crypt_preferred_method could return NULL? I am not against checking it here but hopefully this is an "impossible condition" in libxcrypt.

Reply:
> In which case crypt_preferred_method could return NULL? I am not against checking it here but hopefully this is an "impossible condition" in libxcrypt.

libxcrypt needs one of those hashing methods enabled to _NOT_ return `NULL`:

* yescrypt ($y$, strongest)
* bcrypt ($2b$, still considered fairly strong), or
* sha512crypt ($6$, any Linux system *must* support this for compatibility with glibc).


In theory it can return `NULL`, when there is no strong enough hashing method selected during build-time configuration (or in near future by runtime preset), but that would lead to a generally br0k3n Linux system in any case.  For the almost impossible likeliness (or some other libcrypt is used by the system) reason, I've introduced a guaranteed to be invalid `algoid` right below.

Reply:
Well, libxcrypt supports all (except for `crypt16`) of the [crypt() methods](https://en.wikipedia.org/wiki/Crypt_(C)#Support_in_operating_systems) ever used on any system and adds scrypt and (gost-)yescrypt on top of them.

If `glibc` (all distributions that already ship libxcrypt do that) is selected as available hashing group during build-time config, `sha256crypt` will be supported as well, but will not be a preferred method; if just the `glibc` group is selected, `sha512crypt` is preferred.